File: 백준/Gold/1976. 여행 가자/여행 가자.py
import sys
n = int(sys.stdin.readline().rstrip())
m = int(sys.stdin.readline().rstrip())
parent = [i for i in range(n + 1)]


def find_parent(p, x):
    if p[x] != x:
        p[x] = find_parent(p, p[x])
    return p[x]


def union_parent(p, a, b):
    a = find_parent(p, a)
    b = find_parent(p, b)
    if a < b:
        p[b] = a
    elif a > b:
        p[a] = b
    else:
        return


g = [[0 for _ in range(n + 1)] for __ in range(n + 1)]
for i in range(1, n + 1):
    g[i] = list(map(int, sys.stdin.readline().rstrip().split()))
    g[i].insert(0, 0)
    for j in range(1, n + 1):
        if g[i][j] and find_parent(parent, i) != find_parent(parent, j):
            union_parent(parent, i, j)
travel = list(map(int, sys.stdin.readline().rstrip().split()))
ans_list = []
for t in travel:
    ans_list.append(parent[t])
if len(set(ans_list)) == 1:
    print("YES")
else:
    print("NO")

File: 백준/Gold/1976. 여행 가자/여행 가자.cc
#include <iostream>
#include <set>
using namespace std;
int n, m;
int g[202][202];
int parent[202];
int find_parent(int parent[], int x) {
	if (parent[x] == x) return x;
	return parent[x] = find_parent(parent, parent[x]);
}
void union_parent(int parent[], int a, int b) {
	a = find_parent(parent, a);
	b = find_parent(parent, b);
	if (a < b)		parent[b] = a;
	else if (a > b) parent[a] = b;
	else			return;
}

int main() {
	ios_base::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);
	cin >> n >> m;
	for (int i = 1; i < n + 1; i++)
		parent[i] = i;
	for (int i = 1; i < n + 1; i++) {
		for (int j = 1; j < n + 1; j++) {
			cin >> g[i][j];
			if (g[i][j] == 1 && find_parent(parent, i) != find_parent(parent, j))
				union_parent(parent, i, j);
		}
	}
	set<int> s;
	int travel;
	for (int i = 0; i < m; i++) {
		cin >> travel;
		s.insert(parent[travel]);
	}
	if (s.size() == 1)	cout << "YES" << endl;
	else				cout << "NO" << endl;
	return 0;
}